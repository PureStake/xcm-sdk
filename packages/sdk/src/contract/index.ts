export * from './contract.factory';
export * from './contract.interfaces';
export * from './contracts';
