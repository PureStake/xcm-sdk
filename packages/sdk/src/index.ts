export * from './getSourceData';
export * from './sdk';
export * from './sdk.interfaces';
