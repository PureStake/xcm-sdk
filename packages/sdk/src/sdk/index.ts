export * from './sdk';
export * from './sdk.interfaces';
export { isXcmSdkDeposit, isXcmSdkWithdraw, toDecimal } from './sdk.utils';
