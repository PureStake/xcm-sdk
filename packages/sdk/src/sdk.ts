/* eslint-disable sort-keys */
import {
  ConfigBuilder,
  ConfigService,
  IConfigService,
} from '@moonbeam-network/xcm-config';
import {
  AnyChain,
  Asset,
  AssetAmount,
  Ecosystem,
} from '@moonbeam-network/xcm-types';
import { getAssetsBalances } from './getTransferData/getSourceData';
import { getTransferData as gtd } from './getTransferData/getTransferData';
import { PolkadotService } from './polkadot';
import { Signers, TransferData } from './sdk.interfaces';

export interface SdkOptions extends Partial<Signers> {
  configService?: IConfigService;
}

export function Sdk(options?: SdkOptions) {
  const configService = options?.configService ?? new ConfigService();

  return {
    assets(ecosystem?: Ecosystem) {
      const { assets, asset } = ConfigBuilder(configService).assets(ecosystem);

      return {
        assets,
        asset(keyOrAsset: string | Asset) {
          const { sourceChains, source } = asset(keyOrAsset);

          return {
            sourceChains,
            source(keyOrChain: string | AnyChain) {
              const { destinationChains, destination } = source(keyOrChain);

              return {
                destinationChains,
                destination(destKeyOrChain: string | AnyChain) {
                  return {
                    async accounts(
                      sourceAddress: string,
                      destinationAddress: string,
                    ): Promise<TransferData> {
                      return gtd({
                        ...options,
                        configService,
                        destinationAddress,
                        sourceAddress,
                        transferConfig: destination(destKeyOrChain).build(),
                      });
                    },
                  };
                },
              };
            },
          };
        },
      };
    },
    async getTransferData({
      destinationAddress,
      destinationKeyOrChain,
      evmSigner,
      keyOrAsset,
      polkadotSigner,
      sourceAddress,
      sourceKeyOrChain,
    }: SdkTransferParams): Promise<TransferData> {
      return gtd({
        configService,
        destinationAddress,
        evmSigner,
        polkadotSigner,
        sourceAddress,
        transferConfig: ConfigBuilder(configService)
          .assets()
          .asset(keyOrAsset)
          .source(sourceKeyOrChain)
          .destination(destinationKeyOrChain)
          .build(),
      });
    },
  };
}

export async function getParachainBalances(
  chain: AnyChain,
  address: string,
): Promise<AssetAmount[]> {
  const configService = new ConfigService();
  const chainsConfig = configService.getChainConfig(chain);
  const assets = chainsConfig.getAssetsConfigs();

  const polkadot = await PolkadotService.create(chain, configService);

  const balances = await getAssetsBalances({
    chain,
    assets,
    address,
    polkadot,
  });

  return balances;
}

export interface SdkTransferParams extends Partial<Signers> {
  destinationAddress: string;
  destinationKeyOrChain: string | AnyChain;
  keyOrAsset: string | Asset;
  sourceAddress: string;
  sourceKeyOrChain: string | AnyChain;
}
