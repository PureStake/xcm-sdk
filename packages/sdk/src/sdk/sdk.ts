/* eslint-disable @typescript-eslint/no-use-before-define */
import {
  Asset,
  AssetSymbol,
  Chain,
  ChainKey,
  moonbase,
  MoonbaseAssets,
  MoonbaseChains,
  moonbeam,
  MoonbeamAssets,
  MoonbeamChains,
  moonriver,
  MoonriverAssets,
  MoonriverChains,
  XcmConfigBuilder,
} from '@moonbeam-network/xcm-config';
import { UnsubscribePromise } from '@polkadot/api/types';
import { IKeyringPair } from '@polkadot/types/types';
import { isString } from 'lodash';
import { XTokensContract } from '../contracts';
import { AssetBalanceInfo, createPolkadotServices } from '../polkadot';
import { getDepositData } from './sdk.deposit';
import {
  DepositTransferData,
  SdkOptions,
  WithdrawGetParams,
  WithdrawTransferData,
  XcmSdk,
  XcmSdkByChain,
  XcmSdkDeposit,
  XcmSdkWithdraw,
} from './sdk.interfaces';
import { createDummyAsset, subscribeToAssetsBalanceInfo } from './sdk.utils';
import { getWithdrawData } from './sdk.withdraw';

export function init(options?: SdkOptions): XcmSdkByChain {
  return {
    moonbase: initByChain<MoonbaseAssets, MoonbaseChains>(moonbase, options),
    moonbeam: initByChain<MoonbeamAssets, MoonbeamChains>(moonbeam, options),
    moonriver: initByChain<MoonriverAssets, MoonriverChains>(
      moonriver,
      options,
    ),
  };
}

function initByChain<Symbols extends AssetSymbol, ChainKeys extends ChainKey>(
  configBuilder: XcmConfigBuilder<Symbols, ChainKeys>,
  options?: SdkOptions,
): XcmSdk<Symbols, ChainKeys> {
  const { symbols, assets, moonAsset, moonChain } = configBuilder;

  return {
    symbols,
    assets,
    moonAsset,
    moonChain,
    subscribeToAssetsBalanceInfo: async (
      account: string,
      cb: (data: AssetBalanceInfo<Symbols>[]) => void,
    ): UnsubscribePromise => {
      const [polkadot] = await createPolkadotServices<Symbols, ChainKeys>([
        configBuilder.moonChain.ws,
      ]);

      return subscribeToAssetsBalanceInfo({
        account,
        polkadot,
        configBuilder,
        cb,
      });
    },
    deposit: (
      symbolOrAsset: Symbols | Asset<Symbols>,
    ): XcmSdkDeposit<Symbols, ChainKeys> => {
      const { chains, from } = configBuilder.deposit(symbolOrAsset);

      return {
        chains,
        from: (keyOrChain: ChainKeys | Chain<ChainKeys>) => {
          const { asset, origin, config } = from(keyOrChain);

          return {
            get: async (
              account: string,
              sourceAccount: string | IKeyringPair,
              { primaryAccount, polkadotSigner } = {},
            ): Promise<DepositTransferData<Symbols>> => {
              const signer = polkadotSigner || options?.polkadotSigner;

              if (isString(sourceAccount) && !signer) {
                throw new Error(
                  'Polkadot Signer/KeyringPair are not provided to XCM-SDK',
                );
              }

              const [polkadot, foreignPolkadot] = await createPolkadotServices<
                Symbols,
                ChainKeys
              >([configBuilder.moonChain.ws, config.origin.ws]);
              const meta = foreignPolkadot.getMetadata();
              const nativeAsset =
                configBuilder.assets[meta.symbol] ||
                createDummyAsset(meta.symbol);

              return getDepositData({
                account,
                asset,
                config,
                foreignPolkadot,
                moonChain,
                nativeAsset,
                origin,
                polkadot,
                polkadotSigner: signer,
                primaryAccount,
                sourceAccount,
              });
            },
          };
        },
      };
    },
    withdraw: (
      symbolOrAsset: Symbols | Asset<Symbols>,
    ): XcmSdkWithdraw<Symbols, ChainKeys> => {
      const { chains, to } = configBuilder.withdraw(symbolOrAsset);

      return {
        chains,
        to: (keyOrChain: ChainKeys | Chain<ChainKeys>) => {
          const { asset, origin, config } = to(keyOrChain);

          return {
            get: async (
              destinationAccount: string,
              { ethersSigner }: WithdrawGetParams = {},
            ): Promise<WithdrawTransferData<Symbols>> => {
              const signer = ethersSigner || options?.ethersSigner;

              if (!signer) {
                throw new Error('Ethers signer is not provided to XCM-SDK');
              }

              const contract = new XTokensContract<Symbols>(signer);
              const [polkadot, destinationPolkadot] =
                await createPolkadotServices<Symbols, ChainKeys>([
                  configBuilder.moonChain.ws,
                  config.destination.ws,
                ]);
              const meta = destinationPolkadot.getMetadata();
              const nativeAsset = configBuilder.assets[meta.symbol];

              return getWithdrawData({
                asset,
                config,
                contract,
                destinationAccount,
                destinationPolkadot,
                ethersSigner: signer,
                moonChain,
                nativeAsset,
                origin,
                polkadot,
              });
            },
          };
        },
      };
    },
  };
}
