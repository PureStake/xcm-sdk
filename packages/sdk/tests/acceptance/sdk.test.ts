/* eslint-disable jest/no-commented-out-tests */
import {
  centrifuge,
  hydraDX,
  hydraDxAlphanet,
  moonbaseAlpha,
  moonbaseBeta,
  moonbeam,
} from '@moonbeam-network/xcm-config';
import { getParachainBalances } from '../../src/sdk';

// E2E balance test wallet
const hydraDXAddress = '7MR8Qxy9sJmN6bfHMggAtFY5DwLxfrssLuTnP5rmkpD92oPH';
const centrifugeAddress = '4fAKSBMGVT9jt1jkuJvXgvMbmqV2BuspFWWEmdVeFj9yRudb';
const moonbeamAddress = '0x4E82143Af671Cc8201Bc7efCBbCED3A69e84405e';
const substrateAddress = '5FtGz8bgoCQ6pNAYLWCfxKx9ekLnX1ewP9q2TjMT2riu7sf9';
const moonbaseBetaAddress = '0x08480769599E23F626efff39B89F3137e9917a40';

describe('sdk', () => {
  describe(`${getParachainBalances.name}`, () => {
    it(`should get expected balances for ${moonbeam.name}`, async () => {
      const result = await getParachainBalances(moonbeam, moonbeamAddress);

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${hydraDX.name}`, async () => {
      const result = await getParachainBalances(hydraDX, hydraDXAddress);

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${hydraDX.name} for substrate address`, async () => {
      const result = await getParachainBalances(hydraDX, substrateAddress);

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${centrifuge.name}`, async () => {
      const result = await getParachainBalances(centrifuge, centrifugeAddress);

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${centrifuge.name}  for substrate address`, async () => {
      const result = await getParachainBalances(centrifuge, substrateAddress);

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${hydraDxAlphanet.name}`, async () => {
      const result = await getParachainBalances(
        hydraDxAlphanet,
        hydraDXAddress,
      );

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${hydraDxAlphanet.name} for substrate address`, async () => {
      const result = await getParachainBalances(
        hydraDxAlphanet,
        substrateAddress,
      );

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${moonbaseBeta.name}`, async () => {
      const result = await getParachainBalances(
        moonbaseBeta,
        moonbaseBetaAddress,
      );

      expect(result).toMatchSnapshot();
    });
    it(`should get expected balances for ${moonbaseAlpha.name}`, async () => {
      const result = await getParachainBalances(moonbaseAlpha, moonbeamAddress);

      expect(result).toMatchSnapshot();
    });
  });
});

/* eslint-disable jest/max-expects */
/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable jest/no-standalone-expect */
/* eslint-disable jest/require-hook */
// import {
//   AssetSymbol,
//   ChainKey,
//   MoonChainKey,
// } from '@moonbeam-network/xcm-config';
// import { KeyringPair } from '@polkadot/keyring/types';
// import { Wallet } from 'ethers';
// import _ from 'lodash';
// import { ExtrinsicStatus, init } from '../../src';
// import { getEthersWalletSigner, getPolkadotKeyringPair } from '../utils/auth';
// import { itIf } from '../utils/itIf';

// const CONFIG = {
//   [MoonChainKey.Moonbeam]: {
//     [AssetSymbol.IBTC]: {
//       deposit: { [ChainKey.Interlay]: true },
//       withdraw: { [ChainKey.Interlay]: true },
//     },
//     [AssetSymbol.USDT]: {
//       deposit: { [ChainKey.polkadotAssetHub]: true },
//       withdraw: { [ChainKey.polkadotAssetHub]: true },
//     },
//   },
// };

// describe('sdk', () => {
//   const { moonbeam } = init();
//   // to hide polkadot unnecessary warnings
//   const consoleWarnMock = jest.spyOn(console, 'warn').mockImplementation();

//   let ethersSigner: Wallet;
//   let keyringPair: KeyringPair;

//   beforeAll(async () => {
//     ethersSigner = getEthersWalletSigner(moonbeam.moonChain);
//     keyringPair = await getPolkadotKeyringPair();
//   });

//   afterAll(() => consoleWarnMock.mockRestore());

//   describe.each([moonbeam])('on $moonChain.name', (sdk) => {
//     describe.each(sdk.symbols)('%s', (symbol) => {
//       const { chains: fromChains, from } = sdk.deposit(symbol);
//       const { chains: toChains, to } = sdk.withdraw(symbol);

//       fromChains.length &&
//         describe.each(fromChains)('deposit from $key', (chain) => {
//           const skip: boolean | undefined = _.get(CONFIG, [
//             sdk.moonChain.key,
//             symbol,
//             'deposit',
//             chain.key,
//           ]);

//           itIf(!skip)('should deposit', (done) => {
//             from(chain)
//               .get(ethersSigner.address, keyringPair)
//               .then(({ min, send }) => {
//                 const hash = send(min, (event) => {
//                   expect(event.status).not.toBe(ExtrinsicStatus.Failed);

//                   if (event.status === ExtrinsicStatus.Success) {
//                     expect(event.txHash).toBeDefined();
//                     expect(event.blockHash).toBeDefined();
//                     done();
//                   }
//                 });

//                 expect(hash).toBeDefined();
//               });
//           });
//         });

//       toChains.length &&
//         describe.each(toChains)('withdraw to $key', (chain) => {
//           const skip: boolean | undefined = _.get(CONFIG, [
//             sdk.moonChain.key,
//             symbol,
//             'withdraw',
//             chain.key,
//           ]);

//           itIf(!skip)('should withdraw', (done) => {
//             to(chain)
//               .get(keyringPair.address, {
//                 ethersSigner,
//               })
//               .then(({ min, send }) => {
//                 const hash = send(min, (event) => {
//                   expect(event.status).not.toBe(ExtrinsicStatus.Failed);

//                   if (event.status === ExtrinsicStatus.Success) {
//                     expect(event.txHash).toBeDefined();
//                     expect(event.blockHash).toBeDefined();
//                     done();
//                   }
//                 });

//                 expect(hash).toBeDefined();
//               });
//           });
//         });
//     });
//   });
// });
