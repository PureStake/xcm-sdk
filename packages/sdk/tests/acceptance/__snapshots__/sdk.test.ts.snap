// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sdk > getParachainBalances > on 'Centrifuge' for address: '4fAKSBMGVT9jt1jkuJvXgvMbmqV2BuspFWWEm…' > should get expected balances 1`] = `
[
  a {
    "amount": 570000000000000000n,
    "decimals": 18,
    "key": "cfg",
    "originSymbol": "CFG",
    "symbol": "CFG",
  },
]
`;

exports[`sdk > getParachainBalances > on 'Centrifuge' for address: '5FtGz8bgoCQ6pNAYLWCfxKx9ekLnX1ewP9q2T…' > should get expected balances 1`] = `
[
  a {
    "amount": 570000000000000000n,
    "decimals": 18,
    "key": "cfg",
    "originSymbol": "CFG",
    "symbol": "CFG",
  },
]
`;

exports[`sdk > getParachainBalances > on 'HydraDX Alphanet' for address: '5FtGz8bgoCQ6pNAYLWCfxKx9ekLnX1ewP9q2T…' > should get expected balances 1`] = `
[
  a {
    "amount": 178046239082538836n,
    "decimals": 18,
    "key": "ftmwh",
    "originSymbol": "FTM.wh",
    "symbol": "FTM.wh",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dev",
    "originSymbol": "DEV",
    "symbol": "DEV",
  },
  a {
    "amount": 3000000000000n,
    "decimals": 12,
    "key": "hdx",
    "originSymbol": "HDX",
    "symbol": "HDX",
  },
]
`;

exports[`sdk > getParachainBalances > on 'HydraDX Alphanet' for address: '7MR8Qxy9sJmN6bfHMggAtFY5DwLxfrssLuTnP…' > should get expected balances 1`] = `
[
  a {
    "amount": 178046239082538836n,
    "decimals": 18,
    "key": "ftmwh",
    "originSymbol": "FTM.wh",
    "symbol": "FTM.wh",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dev",
    "originSymbol": "DEV",
    "symbol": "DEV",
  },
  a {
    "amount": 3000000000000n,
    "decimals": 12,
    "key": "hdx",
    "originSymbol": "HDX",
    "symbol": "HDX",
  },
]
`;

exports[`sdk > getParachainBalances > on 'HydraDX' for address: '5FtGz8bgoCQ6pNAYLWCfxKx9ekLnX1ewP9q2T…' > should get expected balances 1`] = `
[
  a {
    "amount": 37000000000000n,
    "decimals": 18,
    "key": "weth",
    "originSymbol": "WETH",
    "symbol": "WETH",
  },
  a {
    "amount": 120n,
    "decimals": 8,
    "key": "wbtc",
    "originSymbol": "WBTC",
    "symbol": "WBTC",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdtwh",
    "originSymbol": "USDT.Wh",
    "symbol": "USDT.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dai",
    "originSymbol": "DAI",
    "symbol": "DAI",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "glmr",
    "originSymbol": "GLMR",
    "symbol": "GLMR",
  },
  a {
    "amount": 1000000000000n,
    "decimals": 12,
    "key": "hdx",
    "originSymbol": "HDX",
    "symbol": "HDX",
  },
]
`;

exports[`sdk > getParachainBalances > on 'HydraDX' for address: '7MR8Qxy9sJmN6bfHMggAtFY5DwLxfrssLuTnP…' > should get expected balances 1`] = `
[
  a {
    "amount": 37000000000000n,
    "decimals": 18,
    "key": "weth",
    "originSymbol": "WETH",
    "symbol": "WETH",
  },
  a {
    "amount": 120n,
    "decimals": 8,
    "key": "wbtc",
    "originSymbol": "WBTC",
    "symbol": "WBTC",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdtwh",
    "originSymbol": "USDT.Wh",
    "symbol": "USDT.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dai",
    "originSymbol": "DAI",
    "symbol": "DAI",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "glmr",
    "originSymbol": "GLMR",
    "symbol": "GLMR",
  },
  a {
    "amount": 1000000000000n,
    "decimals": 12,
    "key": "hdx",
    "originSymbol": "HDX",
    "symbol": "HDX",
  },
]
`;

exports[`sdk > getParachainBalances > on 'Moonbase Alpha' for address: '0x4E82143Af671Cc8201Bc7efCBbCED3A69e8…' > should get expected balances 1`] = `
[
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "agng",
    "originSymbol": "AGNG",
    "symbol": "AGNG",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "hdx",
    "originSymbol": "HDX",
    "symbol": "HDX",
  },
  a {
    "amount": 1234567890000000000n,
    "decimals": 18,
    "key": "ftmwh",
    "originSymbol": "FTM.wh",
    "symbol": "FTM.wh",
  },
  a {
    "amount": 10000n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "unit",
    "originSymbol": "Unit",
    "symbol": "Unit",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "tur",
    "originSymbol": "TUR",
    "symbol": "TUR",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "tt1",
    "originSymbol": "TT1",
    "symbol": "TT1",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "pica",
    "originSymbol": "PICA",
    "symbol": "PICA",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "atom",
    "originSymbol": "ATOM",
    "symbol": "ATOM",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "otp",
    "originSymbol": "OTP",
    "symbol": "OTP",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "ampe",
    "originSymbol": "AMPE",
    "symbol": "AMPE",
  },
  a {
    "amount": 10000000000000000n,
    "decimals": 18,
    "key": "alan",
    "originSymbol": "ALAN",
    "symbol": "ALAN",
  },
  a {
    "amount": 6543210000000000000n,
    "decimals": 18,
    "key": "dev",
    "originSymbol": "DEV",
    "symbol": "DEV",
  },
]
`;

exports[`sdk > getParachainBalances > on 'Moonbase Beta' for address: '0x4E82143Af671Cc8201Bc7efCBbCED3A69e8…' > should get expected balances 1`] = `
[
  a {
    "amount": 100000000000000000n,
    "decimals": 18,
    "key": "ftmwh",
    "originSymbol": "FTM.wh",
    "symbol": "FTM.wh",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 1100000000000000000n,
    "decimals": 18,
    "key": "alan",
    "originSymbol": "ALAN",
    "symbol": "ALAN",
  },
  a {
    "amount": 1100000000000000000n,
    "decimals": 18,
    "key": "dev",
    "originSymbol": "DEV",
    "symbol": "DEV",
  },
]
`;

exports[`sdk > getParachainBalances > on 'Moonbeam' for address: '0x4E82143Af671Cc8201Bc7efCBbCED3A69e8…' > should get expected balances 1`] = `
[
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "peaq",
    "originSymbol": "PEAQ",
    "symbol": "PEAQ",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "bncs",
    "originSymbol": "BNCS",
    "symbol": "BNCS",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "sub",
    "originSymbol": "SUB",
    "symbol": "SUB",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "ztg",
    "originSymbol": "ZTG",
    "symbol": "ZTG",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "fil",
    "originSymbol": "FIL",
    "symbol": "FIL",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "weth",
    "originSymbol": "WETH",
    "symbol": "WETH",
  },
  a {
    "amount": 0n,
    "decimals": 8,
    "key": "wbtc",
    "originSymbol": "WBTC",
    "symbol": "WBTC",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "vmanta",
    "originSymbol": "vMANTA",
    "symbol": "vMANTA",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "vglmr",
    "originSymbol": "vGLMR",
    "symbol": "vGLMR",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "vfil",
    "originSymbol": "vFIL",
    "symbol": "vFIL",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "vdot",
    "originSymbol": "vDOT",
    "symbol": "vDOT",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "vastr",
    "originSymbol": "vASTR",
    "symbol": "vASTR",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdtwh",
    "originSymbol": "USDT.Wh",
    "symbol": "USDT.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dai",
    "originSymbol": "DAI",
    "symbol": "DAI",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "hdx",
    "originSymbol": "HDX",
    "symbol": "HDX",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "ded",
    "originSymbol": "DED",
    "symbol": "DED",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "pink",
    "originSymbol": "PINK",
    "symbol": "PINK",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdc",
    "originSymbol": "USDC",
    "symbol": "USDC",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdt",
    "originSymbol": "USDT",
    "symbol": "USDT",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "ring",
    "originSymbol": "RING",
    "symbol": "RING",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "pha",
    "originSymbol": "PHA",
    "symbol": "PHA",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "pen",
    "originSymbol": "PEN",
    "symbol": "PEN",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "para",
    "originSymbol": "PARA",
    "symbol": "PARA",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "neuro",
    "originSymbol": "NEURO",
    "symbol": "NEURO",
  },
  a {
    "amount": 0n,
    "decimals": 11,
    "key": "nodl",
    "originSymbol": "NODL",
    "symbol": "NODL",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "manta",
    "originSymbol": "MANTA",
    "symbol": "MANTA",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "ldot",
    "originSymbol": "LDOT",
    "symbol": "LDOT",
  },
  a {
    "amount": 0n,
    "decimals": 10,
    "key": "intr",
    "originSymbol": "INTR",
    "symbol": "INTR",
  },
  a {
    "amount": 0n,
    "decimals": 8,
    "key": "ibtc",
    "originSymbol": "IBTC",
    "symbol": "IBTC",
  },
  a {
    "amount": 0n,
    "decimals": 9,
    "key": "eqd",
    "originSymbol": "EQD",
    "symbol": "EQD",
  },
  a {
    "amount": 0n,
    "decimals": 9,
    "key": "eq",
    "originSymbol": "EQ",
    "symbol": "EQ",
  },
  a {
    "amount": 330000000n,
    "decimals": 10,
    "key": "dot",
    "originSymbol": "DOT",
    "symbol": "DOT",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "cfg",
    "originSymbol": "CFG",
    "symbol": "CFG",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "bnc",
    "originSymbol": "BNC",
    "symbol": "BNC",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "aseed",
    "originSymbol": "aSEED",
    "symbol": "aSEED",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "astr",
    "originSymbol": "ASTR",
    "symbol": "ASTR",
  },
  a {
    "amount": 0n,
    "decimals": 12,
    "key": "aca",
    "originSymbol": "ACA",
    "symbol": "ACA",
  },
  a {
    "amount": 1330000000000000000n,
    "decimals": 18,
    "key": "glmr",
    "originSymbol": "GLMR",
    "symbol": "GLMR",
  },
]
`;

exports[`sdk > getParachainBalances > on 'peaq Alphanet' for address: '5FtGz8bgoCQ6pNAYLWCfxKx9ekLnX1ewP9q2T…' > should get expected balances 1`] = `
[
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "ftmwh",
    "originSymbol": "FTM.wh",
    "symbol": "FTM.wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dev",
    "originSymbol": "DEV",
    "symbol": "DEV",
  },
  a {
    "amount": 1223300000000000000n,
    "decimals": 18,
    "key": "agng",
    "originSymbol": "AGNG",
    "symbol": "AGNG",
  },
]
`;

exports[`sdk > getParachainBalances > on 'peaq EVM Alphanet' for address: '0x4E82143Af671Cc8201Bc7efCBbCED3A69e8…' > should get expected balances 1`] = `
[
  a {
    "amount": 330000000000000000n,
    "decimals": 18,
    "key": "dev",
    "originSymbol": "DEV",
    "symbol": "DEV",
  },
  a {
    "amount": 123000000000000000n,
    "decimals": 18,
    "key": "ftmwh",
    "originSymbol": "FTM.wh",
    "symbol": "FTM.wh",
  },
]
`;

exports[`sdk > getParachainBalances > on 'peaq EVM' for address: '0x4E82143Af671Cc8201Bc7efCBbCED3A69e8…' > should get expected balances 1`] = `
[
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdtwh",
    "originSymbol": "USDT.Wh",
    "symbol": "USDT.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "weth",
    "originSymbol": "WETH",
    "symbol": "WETH",
  },
  a {
    "amount": 0n,
    "decimals": 8,
    "key": "wbtc",
    "originSymbol": "WBTC",
    "symbol": "WBTC",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dai",
    "originSymbol": "DAI",
    "symbol": "DAI",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "glmr",
    "originSymbol": "GLMR",
    "symbol": "GLMR",
  },
]
`;

exports[`sdk > getParachainBalances > on 'peaq' for address: '5FtGz8bgoCQ6pNAYLWCfxKx9ekLnX1ewP9q2T…' > should get expected balances 1`] = `
[
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdtwh",
    "originSymbol": "USDT.Wh",
    "symbol": "USDT.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "weth",
    "originSymbol": "WETH",
    "symbol": "WETH",
  },
  a {
    "amount": 0n,
    "decimals": 8,
    "key": "wbtc",
    "originSymbol": "WBTC",
    "symbol": "WBTC",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "dai",
    "originSymbol": "DAI",
    "symbol": "DAI",
  },
  a {
    "amount": 0n,
    "decimals": 6,
    "key": "usdcwh",
    "originSymbol": "USDC.Wh",
    "symbol": "USDC.Wh",
  },
  a {
    "amount": 0n,
    "decimals": 18,
    "key": "glmr",
    "originSymbol": "GLMR",
    "symbol": "GLMR",
  },
  a {
    "amount": 1313000000000000000n,
    "decimals": 18,
    "key": "peaq",
    "originSymbol": "PEAQ",
    "symbol": "PEAQ",
  },
]
`;
